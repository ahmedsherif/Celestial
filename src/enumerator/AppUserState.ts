enum AppUserState {
	Guest = "Guest",
	User = "User",
}

export { AppUserState };
