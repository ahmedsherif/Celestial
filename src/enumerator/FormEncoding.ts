enum FormEncoding {
	URLEncoded = "application/x-www-form-urlencoded",
	FormData = "multipart/form-data",
	JSON = "application/json",
}

export { FormEncoding };
