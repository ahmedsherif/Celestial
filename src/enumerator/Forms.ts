export enum FormEncoding {
	URLEncoded = "application/x-www-form-urlencoded",
	FormData = "multipart/form-data",
	JSON = "application/json",
}

export enum AccessTokenPlacement {
	headers = "headers",
	body = "body",
}
