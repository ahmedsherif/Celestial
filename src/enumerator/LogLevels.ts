enum LogLevels {
    // Highest to lowest importance
	error = "error",
	warn = "warn",
	info = "info",
	http = "http",
	verbose = "verbose",
	debug = "debug",
	silly = "silly",
}

export { LogLevels };
