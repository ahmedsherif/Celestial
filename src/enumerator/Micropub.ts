enum MicropubQueryType {
	configuration = "config",
	syndicationTargets = "syndicate-to",
	categories = "category",
	source = "source",
}

export { MicropubQueryType };
