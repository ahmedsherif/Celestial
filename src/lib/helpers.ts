import { Request as ExpressRequest } from "express";
import _merge from "lodash.merge";
import { DateTime } from "luxon";

import { APP_TITLE, APP_SUBTITLE } from "../config/constants";
import { AppUserState } from "../enumerator/AppUserState";
import {
	DefaultPageData,
	PostPageData,
	AuthPageData,
	UserPageData,
} from "../interface/PageData";

const getBaseData = (req: ExpressRequest) => {
	return {
		title: APP_TITLE,
		subtitle: APP_SUBTITLE,
		appState: req.session?.appState || AppUserState.Guest,
		user: {
			microformats: {
				name: req.session?.user?.microformats?.name,
				photo: req.session?.user?.microformats?.photo,
			},
			indieauth: {
				identity: req.session?.user?.profileUrl,
			},
			preferences: {
				timezone: req.session?.user?.preferences?.timezone,
				formEncoding: req.session?.user?.preferences?.formEncoding,
			},
		},
	};
};

/**
 * Currently, the app title, subtitle, app state, and user identity is set by default. You may pass these in as well in case you wish to overwrite the defaults.
 */
const pageDataHelper = (
	req: ExpressRequest,
	data?: object
): DefaultPageData | PostPageData | AuthPageData | UserPageData => {
	return _merge(getBaseData(req), data);
};

const postDataHelper = (req: ExpressRequest, data?: object): PostPageData => {
	return _merge(
		getBaseData(req),
		{
			formDefaults: {
				published: {
					date: DateTime.utc()
						.setZone(req.session?.user?.preferences?.timezone)
						.toFormat("yyyy-MM-dd")
						.toString(),
					time: DateTime.utc()
						.setZone(req.session?.user?.preferences?.timezone)
						.toFormat("HH:mm")
						.toString(),
				},
			},
			micropub: {
				"syndicate-to": req.session?.micropub?.["syndicate-to"],
				"media-endpoint": req.session?.micropub?.["media-endpoint"],
			},
		},
		data
	) as PostPageData;
};

/**
 * @param enumerator {any} TypeScript doesn't support specifying a parameter type check for enums. Be wary of what you're passing in!
 */
const enumValuesAsArray = (enumerator: any): Array<string> => {
	return Object.values(enumerator).map((e: any) => e as string);
};

const isObject = (value: any) => {
	return typeof value === "object" && !Array.isArray(value) && value !== null;
};

export { pageDataHelper, postDataHelper, enumValuesAsArray, isObject };
