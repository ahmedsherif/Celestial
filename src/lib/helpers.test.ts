import { pageDataHelper, postDataHelper, enumValuesAsArray } from "./helpers";
import { AppUserState } from "../enumerator/AppUserState";
import { FormEncoding } from "../enumerator/FormEncoding";

describe("Data helpers for Express routes", () => {
	describe("Data helpers for default pages", () => {
		let request;

		beforeEach(() => {
			request = {
				session: {
					appState: AppUserState.User,
					user: {
						profileUrl: "https://example.com",
						microformats: {
							name: "Jane",
							photo: "https://example.com/",
						},
						preferences: {
							timezone: "Asia/Kolkata",
							formEncoding: FormEncoding.URLEncoded,
						},
					},
				},
			};
		});

		test("should return default page data when no extra data provided", () => {
			const pageData = pageDataHelper(request);

			expect(pageData).toMatchSnapshot();
		});

		test("should return merged page data when extra data provided", () => {
			const pageData = pageDataHelper(request, { priority: "high" });

			expect(pageData).toMatchSnapshot();
		});

		test("should overwrite default page data when extra data provided", () => {
			const pageData = pageDataHelper(request, {
				user: { microformats: { name: "John Doe" } },
			});

			expect(pageData).toMatchSnapshot();
		});
	});

	describe("Data helpers for post pages", () => {
		let request;

		beforeAll(() => {
			// Luxon uses Date constructor internally. We mock the current date-time so the snapshots are consistent.
			Date.now = jest.fn(() => 1482363367071);
		});

		beforeEach(() => {
			request = {
				session: {
					appState: AppUserState.User,
					user: {
						profileUrl: "https://example.com",
						microformats: {
							name: "Jane",
							photo: "https://example.com/",
						},
						preferences: {
							timezone: "Asia/Kolkata",
							formEncoding: FormEncoding.URLEncoded,
						},
					},
				},
			};
		});

		test("should return default post data when no extra data provided", () => {
			const postData = postDataHelper(request);

			expect(postData).toMatchSnapshot();
		});

		test("should return merged post data when extra data provided", () => {
			const postData = postDataHelper(request, { priority: "high" });

			expect(postData).toMatchSnapshot();
		});

		test("should overwrite default post data when extra data provided", () => {
			const postData = postDataHelper(request, {
				formDefaults: {
					published: { date: "2020-01-01", time: "10:00" },
				},
			});

			expect(postData).toMatchSnapshot();
		});
	});
});

describe("Miscellaneous utilities", () => {
	test("should return an array of values for an enum with values", () => {
		enum example {
			foo = "bar",
			hello = "world",
		}

		const result = enumValuesAsArray(example);

		expect(result).toStrictEqual(["bar", "world"]);
	});
});
