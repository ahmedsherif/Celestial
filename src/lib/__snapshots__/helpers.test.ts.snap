// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data helpers for Express routes Data helpers for default pages should overwrite default page data when extra data provided 1`] = `
Object {
  "appState": "User",
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "John Doe",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;

exports[`Data helpers for Express routes Data helpers for default pages should return default page data when no extra data provided 1`] = `
Object {
  "appState": "User",
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "Jane",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;

exports[`Data helpers for Express routes Data helpers for default pages should return merged page data when extra data provided 1`] = `
Object {
  "appState": "User",
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "priority": "high",
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "Jane",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;

exports[`Data helpers for Express routes Data helpers for post pages should overwrite default post data when extra data provided 1`] = `
Object {
  "appState": "User",
  "formDefaults": Object {
    "published": Object {
      "date": "2020-01-01",
      "time": "10:00",
    },
  },
  "micropub": Object {
    "media-endpoint": undefined,
    "syndicate-to": undefined,
  },
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "Jane",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;

exports[`Data helpers for Express routes Data helpers for post pages should return default post data when no extra data provided 1`] = `
Object {
  "appState": "User",
  "formDefaults": Object {
    "published": Object {
      "date": "2016-12-22",
      "time": "05:06",
    },
  },
  "micropub": Object {
    "media-endpoint": undefined,
    "syndicate-to": undefined,
  },
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "Jane",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;

exports[`Data helpers for Express routes Data helpers for post pages should return merged post data when extra data provided 1`] = `
Object {
  "appState": "User",
  "formDefaults": Object {
    "published": Object {
      "date": "2016-12-22",
      "time": "05:06",
    },
  },
  "micropub": Object {
    "media-endpoint": undefined,
    "syndicate-to": undefined,
  },
  "postsNavigation": Array [
    Object {
      "endpoint": "/publish/article",
      "icon": "ðŸ“„",
      "name": "Article",
      "type": "article",
    },
    Object {
      "endpoint": "/publish/note",
      "icon": "ðŸ“”",
      "name": "Note",
      "type": "note",
    },
    Object {
      "endpoint": "/publish/reply",
      "icon": "â†ª",
      "name": "Reply",
      "type": "reply",
    },
    Object {
      "endpoint": "/publish/rsvp",
      "icon": "â†ª",
      "name": "RSVP",
      "type": "rsvp",
    },
    Object {
      "endpoint": "/publish/like",
      "icon": "â™¥",
      "name": "Like",
      "type": "like",
    },
    Object {
      "endpoint": "/publish/repost",
      "icon": "â™º",
      "name": "Repost",
      "type": "repost",
    },
    Object {
      "endpoint": "/publish/photo",
      "icon": "ðŸ“·",
      "name": "Photo",
      "type": "photo",
    },
    Object {
      "endpoint": "/publish/video",
      "icon": "ðŸŽ¥",
      "name": "Video",
      "type": "video",
    },
    Object {
      "endpoint": "/publish/event",
      "icon": "ðŸ“…",
      "name": "Event",
      "type": "event",
    },
  ],
  "priority": "high",
  "subtitle": "A micropub client",
  "title": "Celestial",
  "user": Object {
    "indieauth": Object {
      "identity": "https://example.com",
    },
    "microformats": Object {
      "name": "Jane",
      "photo": "https://example.com/",
    },
    "preferences": Object {
      "formEncoding": "application/x-www-form-urlencoded",
      "timezone": "Asia/Kolkata",
    },
  },
  "version": "0.12.1",
}
`;
